crates/stress/src/lib.rs

```diff
@@ -0,0 +1,329 @@
+            .field("thread_pool_size", &self.thread_pool_size)
+            .field("job_count", &self.job_count)
+            .field("run_time_ms", &self.run_time_ms)
+            .field("progress_interval_ms", &self.progress_interval_ms)
+            .finish()
+    }
+}
+
+/// internal job tracking struct
+struct StressJobInfo<J: StressJob> {
+    job_index: usize,
+    job: J,
+}
+
+/// internal stress runner struct
+struct StressRunner<S: StressSuite, J: StressJob> {
+    config: StressRunConfig<S, J>,
+    run_until: std::time::Instant,
+    next_progress: std::time::Instant,
+    thread_pool: Vec<std::thread::JoinHandle<()>>,
+    should_continue: Arc<Mutex<bool>>,
+    job_count: Arc<Mutex<usize>>,
+    job_queue: Arc<Mutex<VecDeque<StressJobInfo<J>>>>,
+    job_last_index: usize,
+    log_recv: crossbeam_channel::Receiver<StressJobLog>,
+    log_send: crossbeam_channel::Sender<StressJobLog>,
+    stats: StressStats,
+}
+
+impl<S: StressSuite, J: StressJob> StressRunner<S, J> {
+    /// private stress runner constructor
+    #[allow(clippy::mutex_atomic)]
+    fn priv_new(config: StressRunConfig<S, J>) -> Self {
+        let (log_send, log_recv) = crossbeam_channel::unbounded();
+        let run_until = std::time::Instant::now()
+            .checked_add(std::time::Duration::from_millis(config.run_time_ms))
+            .unwrap();
+        let next_progress = std::time::Instant::now()
+            .checked_add(std::time::Duration::from_millis(
+                config.progress_interval_ms,
+            ))
+            .unwrap();
+        let mut runner = StressRunner {
+            config,
+            run_until,
+            next_progress,
+            thread_pool: Vec::new(),
+            should_continue: Arc::new(Mutex::new(true)),
+            job_count: Arc::new(Mutex::new(0)),
+            job_queue: Arc::new(Mutex::new(VecDeque::new())),
+            job_last_index: 0,
+            log_recv,
+            log_send,
+            stats: StressStats {
+                master_tick_count: 0,
+                log_stats: HashMap::new(),
+            },
+        };
+        for _ in 0..runner.config.thread_pool_size {
+            runner.create_thread();
+        }
+        runner.config.suite.start();
+        runner
+    }
+
+    /// give the stress runner some processor time
+    fn tick(&mut self) -> bool {
+        if std::time::Instant::now() > self.run_until {
+            *self.should_continue.lock().unwrap() = false;
+            return false;
+        }
+        {
+            let mut cur_job_count = self.job_count.lock().unwrap();
+            while *cur_job_count < self.config.job_count {
+                (*self.job_queue.lock().unwrap()).push_front(StressJobInfo {
+                    job_index: self.job_last_index,
+                    job: (self.config.job_factory)(),
+                });
+                self.job_last_index += 1;
+                *cur_job_count += 1
+            }
+        }
+        for _ in 0..1000 {
```

Comment:
Why 1000? Since the runner.tick is inside a loop itself, would it be enough to just drain the log_recv() and then move on, rather than trying to receive a fixed number of times? Maybe with a small delay and then a retry if the first try_recv() comes up empty?

A comment at least would help. And maybe putting 1000 into a const?

Reply:
lol, originally this did indeed loop until there was nothing to try_recv... but I had a println! in there, and the IO overhead was more that the log generation, so it was generating logs faster than it could print them, so the loop never terminated.

Will add a comment : )